website/upgrade-guides/0-12.html.markdown

```diff
@@ -39,80 +92,341 @@ Examples of these workarounds include:
+configuration (or with Terraform itself) to be able to read the configuration
+and make a good guess as to what it will achieve, and the default connection
+settings feature left an important detail unstated: how do the provisioners
+access the host?
+
+With this in mind, Terraform v0.12 no longer performs any automatic population
+of `connection` blocks. Instead, if you are using any remote provisioners you
+should explicitly set the connection type and the hostname to connect to:
+
+```hcl
+  connection {
+    type = "ssh"
+    host = self.public_ip
+    # ...
+  }
+```
+
+The automatic upgrade tool will detect existing `connection` blocks that are
+lacking these settings within resource types that are known to have previously
+set defaults, and it will write out an expression that approximates whatever
+selection logic the provider was previously doing in its own implementation.
+
+Unfortunately in some cases the provider did not export the result of the
+possibly-rather-complex host selection expression as a single attribute, and so
+for some resource types the generated `host` expression will be quite
+complicated. We recommend reviewing these and replacing them with a simpler
+expression where possible, since you will often know better than Terraform does
+which of the instance IP addresses are likely to be accessible from the host
+where Terraform is running.
+
+## Upgrades for reusable modules
+
+If you are making upgrades to a reusable module that is consumed by many
+different configurations, you may need to take care with the timing of your
+upgrade and of how you publish it.
+
+We strongly recommend using module versioning, either via a Terraform registry
+or via version control arguments in your module source addresses, to pin
+existing references to the old version of the module and then publish the
+upgraded version under a new version number. If you are using semantic
+versioning, such as in a Terraform registry, the updates made by the upgrade
+tool should be considered a breaking change and published as a new major
+version.
+
+The migration tool will automatically add a `>= 0.12.0` Terraform version
+constraint to indicate that the module has been upgraded to use v0.12-only
+features. By using version constraints, users can gradually update their callers
+to use the newly-upgraded version as they begin to use Terraform v0.12 with
+those modules.
+
+For simpler modules it may be possible to carefully adapt them to be both
+0.11 and 0.12 compatible at the same time, by following the upgrade notes in
+earlier sections and avoiding any 0.12-only features. However, for any module
+using a undocumented workarounds for 0.11 limitations it is unlikely to be
+possible to both update it for Terraform v0.12 and retain v0.11 compatibility
+at the same time, because those undocumented workarounds have been replaced
+with new features in Terraform v0.12.
+
+## Map variables no longer merge when overridden
```

Comment:
I feel like an example of the deprecated usage might be helpful here, though not so strongly that I'd argue if you disagree.

Reply:
I did try to write one here originally but an example of this includes a lot of moving parts (configuration example, invocation example, lots of context setting about how to understand those) and it ended up dwarfing what I was trying to say. I may have another attempt at it before final release, or perhaps split this upgrade guide into multiple pages so there's more room to give context on each part.

I'd like to run with this as-is for now and plan to revisit this question in our final edit, if only because I expect folks to be focused mainly on configuration language updates rather than workflow stuff during beta since we're explicitly asking people not to use the beta in their production systems, and so it's most likely to be run manually on folks' local machines rather than in automation.